src/Data/Serialize/Get.hs

```diff
@@ -402,30 +418,93 @@ ensure n0 = n0 `seq` Get $ \ s0 b0 m0 w0 kf ks -> let
                     in ks s b m0 w0 s
                 else getMore n' s0 ss b0 m0 w0 kf ks
 
+negativeIsolation :: Get a
+negativeIsolation = fail "Attempted to isolate a negative number of bytes"
+
+isolationUnderParse :: Get a
+isolationUnderParse =  fail "not all bytes parsed in isolate"
+
+isolationUnderSupply :: Get a
+isolationUnderSupply = failRaw "too few bytes" ["demandInput"]
+
+isolate0 :: Get a -> Get a
+isolate0 parser = do
+  rest <- get
+  cur <- bytesRead
+  put mempty cur
+  a    <- parser
+  put rest cur
+  pure a
+
 -- | Isolate an action to operating within a fixed block of bytes.  The action
 --   is required to consume all the bytes that it is isolated to.
 isolate :: Int -> Get a -> Get a
-isolate n m = do
-  M.when (n < 0) (fail "Attempted to isolate a negative number of bytes")
-  s <- ensure n
-  let (s',rest) = B.splitAt n s
-  cur <- bytesRead
-  put s' cur
-  a    <- m
-  used <- get
-  unless (B.null used) (fail "not all bytes parsed in isolate")
-  put rest (cur + n)
-  return a
+isolate n m
+  | n < 0 = negativeIsolation
+  | n == 0 = isolate0 m
+  | otherwise = do
+      s <- ensure' n
+      let (s',rest) = B.splitAt n s
+      cur <- bytesRead
+      put s' cur
+      a    <- m
+      used <- get
+      unless (B.null used) isolationUnderParse
+      put rest (cur + n)
+      return a
+
+getAtMost :: Int -> Get B.ByteString
+getAtMost n = do
+  (bs, rest) <- B.splitAt n <$> ensure' 1
+  curr <- bytesRead
+  put rest (curr + B.length bs)
+  pure bs
+
+-- | An incremental version of 'isolate', which doesn't try to read the input
+--   into a buffer all at once.
+isolateLazy :: Int -> Get a -> Get a
+isolateLazy n parser
+  | n < 0 = negativeIsolation
+  | n == 0 = isolate0 parser
+isolateLazy n parser = do
+  initialBytesRead <- bytesRead
+  go initialBytesRead . runGetPartial parser =<< getAtMost n
+  where
+    go :: Int -> Result a -> Get a
+    go initialBytesRead r = case r of
+      FailRaw (msg, stack) bs -> bytesRead >>= put bs >> failRaw msg stack
+      Done a bs
+        | otherwise -> do
+            bytesRead' <- bytesRead
+            -- Technically this matches both undersupply, and underparse
+            -- but we throw undersupply to match strict isolation
+            unless (bytesRead' - initialBytesRead == n) isolationUnderSupply
+            unless (B.null bs) isolationUnderParse
+            pure a
+      Partial cont -> do
+        pos <- bytesRead
+        bs <- getAtMost $ n - (pos - initialBytesRead)
+        -- We want to give the inner parser a chance to determine
+        -- output, but if it returns a continuation, we'll throw
+        -- instead of recursing indefinitely
```

Comment:
I'm not sure I understand why you wouldn't recurse here: it seems reasonable to me that at lazily isolated computation could see more than two input blocks. Perhaps giving some sort of control as input would help to guard against a malicious input? If you had a `Maybe Int` argument that when present was the maximum number of blocks that would be accepted before failing unconditionally?

Reply:
So this comment is about the case where `getAtMost` returns zero bytes.

We want to give the isolated parser the chance to succeed or define an error, if possible, but if the inner parser returns a partial when no input was supplied, then we want to error, otherwise we'll enter a loop of calling `getAtMost` over and over, getting zero bytes.

I think the only time this can happen is when `getAtMost` is called with `0`, so the inner parser *should* finish.

Reply:
Could you modify the comment to include your reasoning about the case where `n - (pos - initialBytesRead)` is zero?

Reply:
Sure!